lib/upgrade/etcd/conversionv1v3/rule.go

```diff
@@ -0,0 +1,266 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package conversionv1v3
+
+import (
+	"fmt"
+	"sync"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+
+	apiv1 "github.com/projectcalico/libcalico-go/lib/apis/v1"
+	apiv3 "github.com/projectcalico/libcalico-go/lib/apis/v3"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	"github.com/projectcalico/libcalico-go/lib/net"
+	"github.com/projectcalico/libcalico-go/lib/numorstring"
+)
+
+// rulesAPIToBackend converts an API Rule structure slice to a Backend Rule structure slice.
+func rulesAPIToBackend(ars []apiv1.Rule) []model.Rule {
+	if ars == nil {
+		return []model.Rule{}
```

Comment:
this is how all our v1 backend values were stored in the backend (empty value, instead of nil), maybe it had something to do with how v1 etcd backend could unmarshal it?

Reply:
ok, happy to keep, but keep an eye out for it.